master/internal/templates/api.go

```diff
@@ -0,0 +1,259 @@
+func (a *TemplateAPIServer) GetTemplates(
+	ctx context.Context,
+	req *apiv1.GetTemplatesRequest,
+) (*apiv1.GetTemplatesResponse, error) {
+	user, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get the user: %s", err)
+	}
+
+	scopes, err := AuthZProvider.Get().ViewableScopes(ctx, user, 0)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to check for permissions: %s", err)
+	}
+
+	var resp apiv1.GetTemplatesResponse
+	if err := db.Bun().NewRaw(`
+SELECT name, config, workspace_id
+FROM templates
+`).Scan(ctx, &resp.Templates); err != nil {
+		return nil, fmt.Errorf("fetching templates from database: %w", err)
+	}
+	api.Where(&resp.Templates, func(i int) bool {
+		if !scopes[model.AccessScopeID(resp.Templates[i].WorkspaceId)] {
+			return false
+		}
+		return strings.Contains(strings.ToLower(resp.Templates[i].Name), strings.ToLower(req.Name))
+	})
+	api.Sort(resp.Templates, req.OrderBy, req.SortBy, apiv1.GetTemplatesRequest_SORT_BY_NAME)
+	return &resp, api.Paginate(&resp.Pagination, &resp.Templates, req.Offset, req.Limit)
+}
+
+// GetTemplate by name. Returns an error if the requested template does not exist.
+func (a *TemplateAPIServer) GetTemplate(
+	ctx context.Context,
+	req *apiv1.GetTemplateRequest,
+) (*apiv1.GetTemplateResponse, error) {
+	user, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "failed to get the user: %s", err)
+	}
+	if len(req.TemplateName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "name is required")
+	}
+
+	var t templatev1.Template
+	err = db.Bun().NewRaw(`
+SELECT name, config, workspace_id
+FROM templates
+WHERE name = ?
+`, req.TemplateName).Scan(ctx, &t)
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, api.NotFoundErrs("template", req.TemplateName, true)
+	case err != nil:
+		return nil, fmt.Errorf("fetching template %s from database: %w", req.TemplateName, err)
+	}
+
+	permErr, err := AuthZProvider.Get().CanViewTemplate(ctx, user, model.AccessScopeID(t.WorkspaceId))
+	switch {
+	case err != nil:
+		return nil, status.Errorf(codes.Internal, "failed to check for permissions: %s", err)
+	case permErr != nil:
+		return nil, authz.SubIfUnauthorized(permErr, api.NotFoundErrs("template", req.TemplateName, true))
+	}
+	return &apiv1.GetTemplateResponse{Template: &t}, nil
+}
+
+// PutTemplate creates or updates a template.
+func (a *TemplateAPIServer) PutTemplate(
+	ctx context.Context,
+	req *apiv1.PutTemplateRequest,
+) (*apiv1.PutTemplateResponse, error) {
+	if len(req.Template.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "name is required")
+	}
+	if req.Template.WorkspaceId != 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "setting workspace_id is not supported.")
+	}
+
+	if _, err := TemplateByName(ctx, req.Template.Name); err != nil {
```

Comment:
should this be ```err is db not found``` instead of ```err != nil```?

Reply:
yep. just copied this from the old code, good catch.